src/day16.scala

```diff
@@ -0,0 +1,201 @@
+// using scala 3.1.0
+package day16
+
+import scala.util.Using
+import scala.io.Source
+import scala.annotation.tailrec
+
+@main def part1(): Unit =
+  println(s"The solution is ${part1(readInput())}")
+
+@main def part2(): Unit =
+  println(s"The solution is ${part2(readInput())}")
+
+def readInput(): String =
+  Using.resource(Source.fromFile("input/day16"))(_.mkString)
+
+val hexadecimalMapping =
+  Map(
+    '0' -> "0000",
+    '1' -> "0001",
+    '2' -> "0010",
+    '3' -> "0011",
+    '4' -> "0100",
+    '5' -> "0101",
+    '6' -> "0110",
+    '7' -> "0111",
+    '8' -> "1000",
+    '9' -> "1001",
+    'A' -> "1010",
+    'B' -> "1011",
+    'C' -> "1100",
+    'D' -> "1101",
+    'E' -> "1110",
+    'F' -> "1111"
+  )
+
+/*
+ * Structures for all possible operators
+ */
+
+sealed trait Packet:
+  def version: Int
+  def typeId: Int
+  def value: Long
+
+sealed trait OperatorPacket extends Packet:
+  def version: Int
+  def typeId: Int
+  def value: Long
+  def exprs: List[Packet]
+
+sealed trait BinaryOperatorPacket extends OperatorPacket:
+  def version: Int
+  def typeId: Int
+  def value: Long
+  def rhs: Packet
+  def lhs: Packet
+  def exprs = List(lhs, rhs)
+
+case class SumPacket(version: Int, exprs: List[Packet]) extends OperatorPacket:
+  val typeId = 0
+  def value = exprs.map(_.value).sum
+
+case class ProductPacket(version: Int, exprs: List[Packet])
+    extends OperatorPacket:
+  val typeId = 1
+  def value = exprs.map(_.value).reduce(_ * _)
+
+case class MinimumPacket(version: Int, exprs: List[Packet])
+    extends OperatorPacket:
+  val typeId = 2
+  def value = exprs.map(_.value).min
+
+case class MaximumPacket(version: Int, exprs: List[Packet])
+    extends OperatorPacket:
+  val typeId = 3
+  def value = exprs.map(_.value).max
+
+case class LiteralPacket(version: Int, value: Long) extends Packet:
+  val typeId = 4
+
+case class GreaterThanPacket(version: Int, lhs: Packet, rhs: Packet)
+    extends BinaryOperatorPacket:
+  val typeId = 5
+  def value = if lhs.value > rhs.value then 1 else 0
+
+case class LesserThanPacket(version: Int, lhs: Packet, rhs: Packet)
+    extends BinaryOperatorPacket:
+  val typeId = 6
+  def value = if lhs.value < rhs.value then 1 else 0
+
+case class EqualsPacket(version: Int, lhs: Packet, rhs: Packet)
+    extends BinaryOperatorPacket:
+  val typeId = 7
+  def value = if lhs.value == rhs.value then 1 else 0
```

Comment:
Maybe? I will take a look.

Reply:
Actually, not sure how to model it, since each case class has different data, the only stable thing is typeId, which I only added for clarity, it's not actually needed.

What would be your idea about it?

Reply:
My first reaction was something similar to

```scala
enum Packet(version: Int, typedId: Int, value: Int):
  case UnaryOperator(version: Int, typedId: Int, value: Int, exprs: List[Packet]) extends Packet(version, typedId, value)
  case BinaryOperator(version: Int, typedId: Int, value: Int, rhs: Packet, lhs: Packet) extends UnaryOperator(version, typedId, value, List(rhs, lhs))
```
although I am not sure it would work for the entire hierarchy

Reply:
What about something like:

```
enum Packet(version: Int):
  case Sum(version: Int, packets: List[Packet]) extends Packet(version)
  case Product(version: Int, packets: List[Packet]) extends Packet(version)
  ...

  def value: Int =
    this match
      case Sum(version, packets) => packets.map(_.value).sum
      case Product(version, packets) => packets.reduce(_ * _)
      ... 
```

Reply:
Changed! Look pretty nice, though summing up versions is a bit more verbose now.